zwave_js_server/client.py

```diff
@@ -199,181 +121,120 @@ async def async_send_json_message(self, message: Dict[str, Any]) -> None:
-                        )
-                    )
-
-            if self.close_requested:
-                break
-
-            self.tries += 1
-
-            try:
-                await self._wait_retry()
-            except asyncio.CancelledError:
-                # Happens if disconnect called
-                break
-
-        self.state = STATE_DISCONNECTED
-        self._disconnect_event.set()
-        self._disconnect_event = None
-
-    async def _wait_retry(self) -> None:
-        """Wait until it's time till the next retry."""
-        # Sleep 2^tries + 0…tries*3 seconds between retries
-        self.retry_task = asyncio.create_task(
-            asyncio.sleep(2 ** min(9, self.tries) + random.randint(0, self.tries * 3))
-        )
-        await self.retry_task
-        self.retry_task = None
+        """Connect to the websocket server."""
+        if self.driver is not None:
+            raise InvalidState("Re-connected with existing driver")
 
-    async def _handle_connection(  # pylint: disable=too-many-branches, too-many-statements
-        self,
-    ) -> None:
-        """Connect to the Z-Wave JS server."""
         client = None
-        disconnect_warn = None
+        self._logger.debug("Trying to connect")
         try:
             self.client = client = await self.aiohttp_session.ws_connect(
                 self.ws_server_url,
                 heartbeat=55,
             )
-            self.tries = 0
 
             self.version = version = VersionInfo.from_message(
                 await client.receive_json()
             )
+        except (
+            client_exceptions.WSServerHandshakeError,
+            client_exceptions.ClientError,
+        ) as err:
+            raise CannotConnect(err) from err
+
+        # basic check for server version compatability
+        self._check_server_version(self.version.server_version)
+
+        self._logger.info(
+            "Connected to Home %s (Server %s, Driver %s)",
+            version.home_id,
+            version.server_version,
+            version.driver_version,
+        )
+        self.state = STATE_CONNECTED
 
-            # basic check for server version compatability
-            self._check_server_version(self.version.server_version)
-
-            self._logger.info(
-                "Connected to Home %s (Server %s, Driver %s)",
-                version.home_id,
-                version.server_version,
-                version.driver_version,
-            )
-            self.state = STATE_CONNECTED
-
-            if self._on_connect:
-                asyncio.create_task(
-                    gather_callbacks(self._logger, "on_connect", self._on_connect)
-                )
-
-            loop = asyncio.get_running_loop()
-
-            while not client.closed:
-                msg = await client.receive()
-
-                if msg.type in (WSMsgType.CLOSED, WSMsgType.CLOSING):
-                    break
+        listen_task = asyncio.create_task(self.listen())
```

Comment:
In your description you mention the consumer has to call `listen` but here you call it inside `connect`. This will make sure `connect` never finishes.

Reply:
That's not good, as we want to wrap connecting + receiving the version message in a timeout.